### ISSUE TITLE
Is it portable?

### ISSUE THREAD

**Issue by ruffianeo** [catid/Zpng#5]
I have a deeply rooted distrust to code which uses raw integer types instead of fixed with types as defined in ``stdint.h``. At best, this way of providing code only causes compiler warnings (Example: the size of the buffer is ``unsigned`` in the code, but should be ``size_t``, which is usually typedef'd as ``unsigned long long``).
My distrust is so deep, that I am compulsed to report the lack of using proper primitive types as an issue.

**Comment by catid** [comment]
Yeah that's a good point I usually don't write APIs that way

**Comment by Clownacy** [comment]
I agree with the usage of `size_t`, but not the fixed-width integers. Those things are arguably less portable than the standard integer types due to being entirely optional - they're not guaranteed by the standard to ever actually exist in the first place. They're also a potential performance sinkhole, forcing the use of integer sizes that are entirely unnatural to the CPU. The whole point of C's standard integer types is that they suit the target CPU, due to not being locked to a particular size across every architecture.

Fixed-width integers only benefit portability when it comes to consistent overflow and underflow behaviour (which can and should be enforced with explicit bit-masking rather than silently relying on an implicit properly of the underlying integer type), and allowing raw data to more-reliably be `memcpy`'d over a struct (which still fails even with fixed-width types when the target platform's word endianness or struct padding rules differ from the one used to generate the raw data). Neither of these behaviours are ones that any good software should rely on, making fixed-width integers nothing more than an unportable crutch for badly-written code.
